File: lib/py/telepact/internal/binary/UnpackBody.py
#|
#|  Copyright The Telepact Authors
#|
#|  Licensed under the Apache License, Version 2.0 (the "License");
#|  you may not use this file except in compliance with the License.
#|  You may obtain a copy of the License at
#|
#|  https://www.apache.org/licenses/LICENSE-2.0
#|
#|  Unless required by applicable law or agreed to in writing, software
#|  distributed under the License is distributed on an "AS IS" BASIS,
#|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#|  See the License for the specific language governing permissions and
#|  limitations under the License.
#|

from msgpack import ExtType

from .PackList import PACKED_BYTE
from .PackMap import UNDEFINED_BYTE


def _unpack_row(row: list[object], header: list[object]) -> dict[object, object]:
    unpacked_row: dict[object, object] = {}
    for index, header_entry in enumerate(header[1:]):
        if index >= len(row):
            break
        value = row[index]
        if type(value) is ExtType and value.code == UNDEFINED_BYTE:
            continue
        if isinstance(header_entry, list):
            nested_key = header_entry[0]
            if value is None:
                unpacked_row[nested_key] = None
                continue
            if type(value) is not list:
                continue
            unpacked_row[nested_key] = _unpack_row(value, header_entry)
            continue
        unpacked_row[header_entry] = value
    return unpacked_row


def _unpack_site(value: object, header: list[object]) -> object:
    if type(value) is not list or len(value) < 2:
        return value
    packed_type = value[0]
    if type(packed_type) is not ExtType or packed_type.code != PACKED_BYTE:
        return value
    if value[1] != header:
        return value
    return [_unpack_row(row, header) for row in value[2:] if type(row) is list]


def unpack_body(body: dict[object, object], binary_encoding: object) -> dict[object, object]:
    unpacked_body = dict(body)
    for path, header in binary_encoding.encoded_pack_sites:
        current: object = unpacked_body
        for key in path[:-1]:
            if type(current) is not dict or key not in current:
                current = None
                break
            current = current[key]
        if type(current) is not dict:
            continue
        final_key = path[-1]
        if final_key not in current:
            continue
        current[final_key] = _unpack_site(current[final_key], header)
    return unpacked_body

File: lib/ts/src/internal/binary/UnpackBody.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { BinaryEncoding, BinaryPackHeader } from './BinaryEncoding.js';
import { MsgpackPacked } from './PackList.js';
import { MsgpackUndefined } from './PackMap.js';

function headersMatch(actual: any, expected: any): boolean {
    if (Array.isArray(actual) && Array.isArray(expected)) {
        if (actual.length !== expected.length) {
            return false;
        }
        for (let index = 0; index < actual.length; index += 1) {
            if (!headersMatch(actual[index], expected[index])) {
                return false;
            }
        }
        return true;
    }
    return actual === expected;
}

function unpackRow(row: any[], header: BinaryPackHeader): Map<any, any> {
    const unpackedRow = new Map<any, any>();
    for (let index = 1; index < header.length; index += 1) {
        if (index - 1 >= row.length) {
            break;
        }
        const value = row[index - 1];
        if (value instanceof MsgpackUndefined) {
            continue;
        }
        const headerEntry = header[index];
        if (Array.isArray(headerEntry)) {
            const nestedKey = headerEntry[0];
            if (value === null) {
                unpackedRow.set(nestedKey, null);
                continue;
            }
            if (!Array.isArray(value)) {
                continue;
            }
            unpackedRow.set(nestedKey, unpackRow(value, headerEntry as BinaryPackHeader));
            continue;
        }
        unpackedRow.set(headerEntry, value);
    }
    return unpackedRow;
}

function unpackSite(value: any, header: BinaryPackHeader): any {
    if (!Array.isArray(value) || value.length < 2 || !(value[0] instanceof MsgpackPacked)) {
        return value;
    }
    if (!headersMatch(value[1], header)) {
        return value;
    }
    return value.slice(2).filter(Array.isArray).map((row) => unpackRow(row, header));
}

export function unpackBody(body: Map<any, any>, binaryEncoding: BinaryEncoding): Map<any, any> {
    const result: Map<any, any> = new Map(body);
    for (const [path, header] of binaryEncoding.encodedPackSites) {
        let current: any = result;
        for (const key of path.slice(0, -1)) {
            if (!(current instanceof Map) || !current.has(key)) {
                current = undefined;
                break;
            }
            current = current.get(key);
        }
        if (!(current instanceof Map)) {
            continue;
        }
        const finalKey = path[path.length - 1]!;
        if (!current.has(finalKey)) {
            continue;
        }
        current.set(finalKey, unpackSite(current.get(finalKey), header));
    }
    return result;
}
